Change the handling of various implementation compatibility options. Most
notably, the "--lint" option will no longer warns about non-POSIX features and
GNU extensions unless POSIX or traditional mode has been explicitly enabled,
and some non-standard features that were accepted when using one of those modes
will now produce fatal errors. No changes to code only affecting "--lint-old"
were made.

--- awkgram.c	2026-07-09 11:48:44.000000000 -0500
+++ awkgram.c	2026-10-28 13:57:15.007376255 -0500
@@ -4273,10 +4273,6 @@
 
 		if (do_traditional || do_posix)
 			yyerror("%s", msg);
-		else if (do_lint_extensions && ! warned) {
-			warned = true;
-			lintwarn("%s", msg);
-		}
 
 		f = yyvsp[0]->lasti;
 		f->opcode = Op_indirect_func_call;
@@ -6370,9 +6366,7 @@
 		pushback();
 		if (c == '[') {
 			if (do_traditional)
-				fatal(_("multidimensional arrays are a gawk extension"));
-			if (do_lint_extensions)
-				lintwarn(_("multidimensional arrays are a gawk extension"));
+				fatal(_("multidimensional arrays are not available in traditional mode"));
 			yylval = GET_INSTRUCTION(Op_sub_array);
 			lasttok = ']';
 		} else {
@@ -6396,8 +6390,6 @@
 			if (nextc(true) == '=') {
 				if (! did_warn_assgn) {
 					did_warn_assgn = true;
-					if (do_lint)
-						lintwarn(_("POSIX does not allow operator `%s'"), "**=");
 					if (do_lint_old)
 						lintwarn(_("operator `%s' is not supported in old awk"), "**=");
 				}
@@ -6407,8 +6399,6 @@
 				pushback();
 				if (! did_warn_op) {
 					did_warn_op = true;
-					if (do_lint)
-						lintwarn(_("POSIX does not allow operator `%s'"), "**");
 					if (do_lint_old)
 						lintwarn(_("operator `%s' is not supported in old awk"), "**");
 				}
@@ -6908,18 +6898,6 @@
 			}
 		}
 
-		if (do_lint) {
-			if (do_lint_extensions && (tokentab[mid].flags & GAWKX) != 0 && (warntab[mid] & GAWKX) == 0) {
-				lintwarn(_("`%s' is a gawk extension"),
-					tokentab[mid].operator);
-				warntab[mid] |= GAWKX;
-			}
-			if ((tokentab[mid].flags & NOT_POSIX) != 0 && (warntab[mid] & NOT_POSIX) == 0) {
-				lintwarn(_("POSIX does not allow `%s'"),
-					tokentab[mid].operator);
-				warntab[mid] |= NOT_POSIX;
-			}
-		}
 		if (do_lint_old && (tokentab[mid].flags & NOT_OLD) != 0
 				 && (warntab[mid] & NOT_OLD) == 0
 		) {
@@ -7246,10 +7224,6 @@
 		(void) mk_rexp(arg);
 
 		if (nexp == 3) {	/* 3rd argument there */
-			if (do_lint_extensions && ! warned) {
-				warned = true;
-				lintwarn(_("match: third argument is a gawk extension"));
-			}
 			if (do_traditional) {
 				yyerror(_("match: third argument is a gawk extension"));
 				return NULL;
@@ -7303,11 +7277,7 @@
 	} else if (r->builtin == do_close) {
 		static bool warned = false;
 		if (nexp == 2) {
-			if (do_lint_extensions && ! warned) {
-				warned = true;
-				lintwarn(_("close: second argument is a gawk extension"));
-			}
-			if (do_traditional) {
+			if (no_gawk_extensions) {
 				yyerror(_("close: second argument is a gawk extension"));
 				return NULL;
 			}
--- awk.h	2026-07-09 11:47:00.000000000 -0500
+++ awk.h	2026-10-28 13:59:17.200385881 -0500
@@ -1248,6 +1248,7 @@
 #define do_debug            (do_flags & DO_DEBUG)
 #define do_mpfr             (do_flags & DO_MPFR)
 #define do_csv              (do_flags & DO_CSV)
+#define no_gawk_extensions  (do_posix | do_traditional)
 
 extern bool do_optimize;
 extern int use_lc_numeric;
--- builtin.c	2026-07-09 11:47:00.000000000 -0500
+++ builtin.c	2026-10-28 13:59:17.200385881 -0500
@@ -587,10 +587,15 @@
 
 	tmp = POP();
 
-	// As of the 2024 POSIX standard, length(array) is standardized. Yay!
+	// Although the 2024 POSIX standard standardizes length(array),
+	// "The Awk Programming Language" by Kernighan et al (1988) only
+	// uses the length function with strings.
 	if (tmp->type == Node_var_array) {
 		unsigned long size;
 
+		if (no_gawk_extensions)
+			fatal(_("length: received array argument"));
+
 		/*
 		 * Support for deferred loading of array elements requires that
 		 * we use the array length interface even though it isn't
--- eval.c	2026-07-09 11:47:00.000000000 -0500
+++ eval.c	2026-10-28 13:57:15.011376288 -0500
@@ -721,7 +721,7 @@
 {
 	static bool warned = false;
 
-	if ((do_lint_extensions || do_traditional) && ! warned) {
+	if (do_lint && no_gawk_extensions && ! warned) {
 		warned = true;
 		lintwarn(_("`IGNORECASE' is a gawk extension"));
 	}
@@ -742,7 +742,7 @@
 	char *p;
 	NODE *v = fixtype(BINMODE_node->var_value);
 
-	if ((do_lint_extensions || do_traditional) && ! warned) {
+	if (do_lint && no_gawk_extensions && ! warned) {
 		warned = true;
 		lintwarn(_("`BINMODE' is a gawk extension"));
 	}
--- field.c	2026-07-09 11:47:00.000000000 -0500
+++ field.c	2026-10-28 13:59:17.200385881 -0500
@@ -1130,8 +1130,6 @@
 	check_args_min_max(nargs, "split", 3, 4);
 
 	if (nargs == 4) {
-		static bool warned = false;
-
 		if (do_traditional) {
 			fatal(_("split: fourth argument is a gawk extension"));
 		}
@@ -1140,10 +1138,6 @@
 			fatal(_("split: fourth argument is not an array"));
 		check_symtab_functab(sep_arr, "split",
 				_("%s: cannot use %s as fourth argument"));
-		if ((do_lint_extensions || do_lint_old) && ! warned) {
-			warned = true;
-			lintwarn(_("split: fourth argument is a gawk extension"));
-		}
 	}
 
 	sep = POP();
@@ -1216,14 +1210,10 @@
 		fs = sep->re_exp;
 
 		if (fs->stlen == 0) {
-			static bool warned = false;
-
 			parseit = null_parse_field;
 
-			if (do_lint && ! warned) {
-				warned = true;
-				lintwarn(_("split: null string for third arg is a non-standard extension"));
-			}
+			if (no_gawk_extensions)
+				fatal(_("split: null string for third arg is a non-standard extension"));
 		} else if ((sep->re_flags & CONSTANT) != 0 || fs->stlen > 1) {
 			// regex constant OR length > 1
 			parseit = re_parse_field;
@@ -1390,11 +1380,11 @@
 	bool fatal_error = false;
 	NODE *tmp;
 
-	if (do_lint_extensions && ! warned) {
+	if (do_lint && ! warned && no_gawk_extensions) {
 		warned = true;
 		lintwarn(_("`FIELDWIDTHS' is a gawk extension"));
 	}
-	if (do_traditional)	/* quick and dirty, does the trick */
+	if (no_gawk_extensions)	/* quick and dirty, does the trick */
 		return;
 
 	/*
@@ -1554,9 +1544,9 @@
 
 		set_parser(null_parse_field);
 
-		if (do_lint_extensions && ! warned) {
+		if (no_gawk_extensions) {
 			warned = true;
-			lintwarn(_("null string for `FS' is a gawk extension"));
+			fatal(_("null string for `FS' is a gawk extension"));
 		}
 	} else if (fs->stlen > 1 || (fs->flags & REGEX) != 0) {
 		if (do_lint_old)
@@ -1688,7 +1678,7 @@
 		warned = true;
 		lintwarn(_("FPAT is a gawk extension"));
 	}
-	if (do_traditional)	/* quick and dirty, does the trick */
+	if (no_gawk_extensions)	/* quick and dirty, does the trick */
 		return;
 
 	/*
--- io.c	2026-07-09 11:47:00.000000000 -0500
+++ io.c	2026-10-28 13:59:17.204385914 -0500
@@ -4214,6 +4214,9 @@
 	save_rs = dupnode(RS_node->var_value);
 	RS_is_null = false;
 	RS = force_string(RS_node->var_value);
+	if (RS->stlen > 1 && no_gawk_extensions) {
+		fatal(_("multicharacter value of `RS' is a gawk extension"));
+	}
 	/*
 	 * used to be if (RS_regexp != NULL) { refree(..); refree(..); ...; }.
 	 * Please do not remerge the if condition; hinders memory deallocation
@@ -4236,19 +4239,12 @@
 		if (first_time || ! do_csv)
 			matchrec = rsnullscan;
 	} else if ((RS->stlen > 1 || (RS->flags & REGEX) != 0) && ! do_posix) {
-		static bool warned = false;
-
 		RS_re[0] = make_regexp(RS->stptr, RS->stlen, false, true, true);
 		RS_re[1] = make_regexp(RS->stptr, RS->stlen, true, true, true);
 		RS_regexp = RS_re[IGNORECASE];
 
 		if (first_time || ! do_csv)
 			matchrec = rsrescan;
-
-		if (do_lint_extensions && ! warned) {
-			lintwarn(_("multicharacter value of RS is a gawk extension"));
-			warned = true;
-		}
 	} else {
 		if (first_time || ! do_csv)
 			matchrec = rs1scan;
--- node.c	2026-07-09 11:47:00.000000000 -0500
+++ node.c	2026-10-28 13:57:15.011376288 -0500
@@ -653,7 +653,7 @@
 		buf[0] = i;
 		break;
 	case 'x':
-		if (do_lint) {
+		if (do_lint && do_posix) {
 			static bool warned = false;
 
 			if (! warned) {
